plugins/catalog-backend/src/service/response/write.ts

```diff
@@ -83,40 +85,59 @@ export async function writeEntitiesResponse(options: {
     const prefix = first ? '[' : ',';
     first = false;
 
-    if (await writeResponseData(res, prefix + entity)) {
+    if ((await writer(prefix + entity)) === 'closed') {
       return;
     }
   }
   res.end(`${first ? '[' : ''}]${trailing}`);
 }
 
 /**
- * Writes a data to the response and waits if the response buffer needs draining.
+ * Creates a data writer that writes to the response and waits if the response
+ * buffer needs draining.
  *
  * @internal
- * @returns true if the response was closed while waiting for the buffer to drain
+ * @returns A writer function. If a write attempt returns 'closed', the
+ * connection has become closed prematurely and the caller should stop trying to
+ * write.
  */
-export async function writeResponseData(res: Response, data: string | Buffer) {
-  const ok = res.write(data, 'utf8');
-  if (!ok) {
+export function createResponseDataWriter(
+  res: Response,
+): (data: string | Buffer) => Promise<'ok' | 'closed'> {
+  // See https://github.com/backstage/backstage/issues/30659
+  //
+  // This code goes to some lengths to just add listeners once at the top,
+  // instead of on every need to drain. Hence it is more complex that seems to
+  // be necessary, just to avoid listener leaks.
+
+  let drainPromise: DeferredPromise<'ok'> | undefined;
+
+  const closePromise = new Promise<'closed'>(resolve => {
+    function onClose() {
+      res.off('drain', onDrain);
+      res.off('close', onClose);
+      res.off('finish', onClose);
+      resolve('closed');
+    }
+    function onDrain() {
+      drainPromise?.resolve('ok');
+      drainPromise = undefined;
+    }
+    res.on('drain', onDrain);
+    res.on('close', onClose);
+    res.on('finish', onClose);
```

Comment:
What's the purpose of also listening to the `finish` event?

Reply:
Hmm, are you sure? I think `close` is emitted both for corner cases and on successful completion